### ISSUE TITLE
Documentation about "how to decorate controllers"

### ISSUE THREAD

**Issue by marsaldev** [PrestaShop/docs#498]
As discussed on #slack, it would be very nice to have this documentation.

The new admin controllers can be decorated?
They are services, so in theory yes, but how it can be done?

Thanks guys 😉

**Comment by mickaelandrieu** [comment]
Hi @marsaldev,

it depends what you means by "decorated" :)

In Symfony a Controller is only about making a matching between HTTP Request conditions and a PHP Callable.

Let's say you have created a Controller named MyController with 3 public function home(), contact(), and search().

You will also have to declare and configure routes (something not required with Legacy Controllers):

```
<?php

namespace FOP\Demo\Controller;

use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\HttpFoundation\Request;

final class MyController extends FrameworkBundleAdminController
{
    /**
     * @Route("/home", methods={"GET", "POST"}, name="fop_demo_home")
     */
    public function home(Request $request)
    {
        return new Response('Hi');
    }

    /* ... */
}
```

Let's say you want in another module (on for a Controller from the Core) override the PHP callable called when matching the HTTP Request conditions, this is what you can do:

In ``your-module-not-demo-one/src/Controller/NameNotImportantController.php

```php
<?php

namespace YourModule\NotDemo\Controller;

use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\HttpFoundation\Request;

final class NameNotImportantController extends FrameworkBundleAdminController
{
    /**
     * @Route("/home", methods={"GET", "POST"}, name="fop_demo_home")
     */
    public function canBeAnotherMethodName(Request $request)
    {
        return new Response('How are you');
    }

    /* ... */
}
```

Note this also can be done when configuring routes using Yaml files.

Hope this helps,

**Comment by marsaldev** [comment]
Hi @mickaelandrieu,
thanks a lot for these infos 😊

I'm talking about the new "core" admin controllers (/src/PrestaShopBundle/Controller/Admin/*), following your tips, for instance, if we want to 'decorate' (but as I see it's like a _substitution_)  the CustomerController (/src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php) we only make a simple route 'override'?

Please get the 'testmodule' attached to see 😊
It doesn't seems works 🤔 I'm missing something?

[testmodule.zip](https://github.com/PrestaShop/docs/files/4438799/testmodule.zip)

**Comment by mickaelandrieu** [comment]
You may need to clear the cache as Yaml file are parsed only once, I don't see any error in your module.

If it doesnt works I'll take a look tonight :)

**Comment by marsaldev** [comment]
> You may need to clear the cache as Yaml file are parsed only once, I don't see any error in your module.

Cleared many times, I get a weird error about another route 🤔


> All Symfony Controllers work the same way, this behavior is also working if you have to override a Controller of a Symfony project using a Bundle :)

This is interesting, and how you tell to the environment how controller should be loaded last? 


> If it doesnt works I'll take a look tonight :)

It should be great, waiting for your check 😉

PS: Tested on 1.7.7.0 version, but I think it's not so different from 1.7.6.x

**Comment by mickaelandrieu** [comment]
> It should be great, waiting for your check :wink:

Looks like I was wrong and it is not *that easy* because the modules are prefixing routes by "modules", just remembered that now :/

Another solution involves to listen  FilterController Event and set the new controller: this requires really good knowledge in Symfony TBH...

> They are services, so in theory yes, but how it can be done?

Are they ? I'm not sure they are in PrestaShop.

If you're able to understand that code I have no worries to have from leaving PrestaShop company :trophy: 

[testmodule.zip](https://github.com/PrestaShop/docs/files/4441312/testmodule.zip)

**Comment by marsaldev** [comment]
I'm not so skilled to find a way with the FilterController 😅, I had a quick chat with Pablo, and I've exposed my thoughts using the official Symfony documentation, and it's not so clear if we want to 'decorate' it. 
As you says it's more simple, in a Symfony native project simply 'override' the route for that action and it's done, but obviously we doesn't like the easy things 🤣 

From the Symfony documentation:
“In Symfony, a controller does not need to be registered as a service. But if you’re using the [default services.yml configuration](https://symfony.com/doc/3.4/service_container.html#service-container-services-load-example), your controllers are already registered as services. This means you can use dependency injection like any other normal service.”
https://symfony.com/doc/3.4/controller/service.html
It talks about DI not decorating 🤔

@matks we need you 😊

**Comment by matks** [comment]
> From the Symfony documentation:
> “In Symfony, a controller does not need to be registered as a service. But if you’re using the [default services.yml configuration](https://symfony.com/doc/3.4/service_container.html#service-container-services-load-example), your controllers are already registered as services. This means you can use dependency injection like any other normal service.”
> https://symfony.com/doc/3.4/controller/service.html

We are not using the default services configuration so our controllers are not registered as services.
Pablo asked me to check it, this is something easy to fix. I did a PR https://github.com/PrestaShop/PrestaShop/pull/18475

**Comment by marsaldev** [comment]
Thanks @matks !

**Comment by matks** [comment]
https://github.com/PrestaShop/PrestaShop/pull/18475/ has been merged
I will submit a PR to explain how to use them 😉

**Comment by marsaldev** [comment]
What a great news! Thanks a lot 😉

**Comment by matks** [comment]
So we found new issues 😅 and tough ones as they seem to come from Symfony.
We opened a bug report to Symfony https://github.com/symfony/symfony/issues/36567

**Comment by matks** [comment]
We are exploring a fix https://github.com/PrestaShop/PrestaShop/pull/18811/

**Comment by marsaldev** [comment]
Yes, I'm watching you 👀🤣

**Comment by matks** [comment]
The PR that should fix this is waiting for QA now https://github.com/PrestaShop/PrestaShop/pull/18811

**Comment by marsaldev** [comment]
Great! 😉

**Comment by matks** [comment]
@marsaldev PR https://github.com/PrestaShop/PrestaShop/pull/18811 has been merged and I have provided a test module to demonstrate it can be done 😄

To complete this topic, I can write a small example for documentation of "how to override a Symfony page : override template OR override routing OR override controller OR (new in 1.7.7) decorate controller

> Are they ? I'm not sure they are in PrestaShop.

@mickaelandrieu Now they are 😉in PS 1.7.7

**Comment by marsaldev** [comment]
What a wonderful news! 🤩
Thanks a lot @matks!

**Comment by matks** [comment]
WIP https://github.com/PrestaShop/docs/pull/546

**Comment by matks** [comment]
https://github.com/PrestaShop/docs/pull/546 is finally merged @marsaldev
